WIP: vt6/term and vt6/mono modules

majewsky opened the pull request:
This is based on an earlier draft (#14) where monospace layouting was a capability within vt6/term. I chose to make an entirely new branch because so much has changed in vt6/core1.0 recently that stuff is way cleaner this way.

@laerling FYI - If you want to read this, start with `term1.0` and then continue with `mono1.0`. The latter depends on the former.

majewsky commented on spec/term/1.0.md:
```diff
@@ -0,0 +1,123 @@
+
+### 1.2. Standard output
+
+When a client writes a byte string into its standard output, the server SHALL assume that this byte string is encoded in [UTF-8](https://tools.ietf.org/html/rfc3629), and decode it into a string of Unicode characters.
```
> When a client writes a byte string into its standard output

That stdout may not be connected to the terminal.

laerling commented on spec/term/1.0.md:
```diff
@@ -0,0 +1,204 @@
+## 1. Scope
+
+This module defines how a terminal works, how clients' standard output is reflected on the terminal, and how user input is presented on clients' standard input.
+Besides [`vt6/core`](https://vt6.io/std/core/), this is the only module that we expect to be available in **every** VT6 server implementation across the entire feature spectrum, from graphical terminal emulators all the way down to continuous stationery printers.
```
term module: Can we safely allow vt6 client implementers to assume the term module to always be present? Or should we advise them to always `want` it?

laerling commented on spec/term/1.0.md:
```diff
@@ -0,0 +1,204 @@
+
+If the server takes text input from a user or some other source, it MAY pass this input to the clients as described in section 1.3.
+
+*Rationale:* This is a "MAY" because some servers (e.g. continuous stationery printers) do not have an input method.
```
Nitpick: That's what the "If" is for. I don't know why there is a "MAY". I SHOULD be a "MUST" ;)

majewsky replied:
No. Not all text input that a terminal takes from a user must be passed on to clients' standard inputs.

laerling replied:
Alright, but then that is the reason which should appear in the text, not "some servers do not have an input method".

laerling commented on spec/term/1.0.md:
```diff
@@ -0,0 +1,204 @@
+Invalid byte sequences or invalid code points SHALL be handled gracefully, for instance, by replacing them with the replacement character (U+FFFD).
+
+When the `term.output-protected` property is true, the server MUST then discard all characters from this string that are in the general categories Cc ("Other, control"), Cf ("Other, format"), Cs ("Other, surrogate"), Co ("Other, private use") and Cn ("Other, not assigned"), except for U+0009 (Horizontal Tab), U+000A (New Line), U+000C (Form Feed) and U+000D (Carriage Return).
+These characters are called **active characters** for the remainder of this specification.
```
Not entirely clear whether "These characters" refers to "all characters from this string that are in the general categories" or "except for".

laerling commented on spec/term/1.0.md:
```diff
@@ -0,0 +1,204 @@
+
+The remaining string of Unicode characters SHALL then be inserted into the terminal document at the position pointed to by the cursor.
+The cursor SHALL be updated to point at the position directly after the inserted string.
+If the terminal document would grow too large to fit within the program's memory because of this operation, the server MAY choose to discard some of the text at the start of the document.
```
Practically the same as in 1.5.

majewsky replied:
Fixed by folding 1.5 into this part.

laerling commented on spec/term/1.0.md:
```diff
@@ -0,0 +1,204 @@
+## 1. Scope
+
+This module defines how a terminal works, how clients' standard output is reflected on the terminal, and how user input is presented on clients' standard input.
+Besides [`vt6/core`](https://vt6.io/std/core/), this is the only module that we expect to be available in **every** VT6 server implementation across the entire feature spectrum, from graphical terminal emulators all the way down to continuous stationery printers.
```
Except muxers and proxies...

majewsky replied:
Muxers and proxies still need to implement `term`. They *are not* the terminal, but they have an internal model of a terminal that they present towards their own clients. A muxer/proxy that only has `core` is less than useless.

laerling replied:
So proxies and muxers would just forward all term-related messages from the terminal to the client and not modify it. Makes sense.

majewsky replied:
Well, not "not modify it". The only modification they do is multiplexing it onto a single connection.

laerling replied:
That's what I meant.

laerling commented on spec/term/1.0.md:
```diff
@@ -0,0 +1,204 @@
+
+## 2. Message types for `vt6/term1`
+
+### 2.1. The `term.make-stdio` message
```
`make-stdio` and `is-stdio` should not be part of `term`, but of `core`, since muxers and proxies must also be able to turn a connection into stdio.

majewsky replied:
See above.

laerling replied:
Let me rephrase: `make-stdio` and `is-stdio` should not be part of `term`, but of `core`, since those are functionalities that *every* server must support, regardless of whether it `have`s `term` or not.
I know, all servers will `have term`, but the point is that the `make-stdio` and `is-stdio` are independent of that.

We must not define things that every server must have inside a non-`core` module that every server must have, when those things don't have anything to do with it semantically!

majewsky replied:
Agreed. Move this to `core` to enforce a logical separation between `core` (which corresponds roughly to OSI layers 1-4) and `term` (which corresponds roughly to OSI layers 5-7).

laerling commented on spec/term/1.0.md:
```diff
@@ -0,0 +1,204 @@
+
+TODO explain in detail (how stdin works, how stdout works, how anything that might send a server-initiated message is cancelled)
+
+TODO maybe remove the notion of standard input/output from core1.0 entirely and introduce it here?
```
See comment above.

majewsky replied:
See rebuttal above.

laerling replied:
See my ranty response above.

laerling commented on spec/term/1.0.md:
```diff
@@ -0,0 +1,204 @@
+- Acceptable values: unsigned integers
+- Can be set by client: no
+
+The value of this property is the maximum number of printable ASCII characters (Unicode codepoints U+0020-U+007F) that can be displayed on the terminal viewport without having to break the line or overflow the viewport.
```
I don't see how this makes any sense outside of mono.

majewsky replied:
OK.

laerling commented on spec/term/1.0.md:
```diff
@@ -0,0 +1,204 @@
+*Rationale:* This is a "MAY" because servers may employ more sophisticated computations to account for ligatures, or simpler computations if they use monospace fonts.
+
+The value of this property SHALL NOT be zero.
+If the computation above would result in a value of zero, the server SHALL report a value of one SHALL instead.
```
What is a SHALL? I suppose you mean "the value of one"?

laerling commented on spec/term/1.0.md:
```diff
@@ -0,0 +1,204 @@
+### 3.8. The `term.output-wordwrap` property
+
+- Acceptable values: booleans
+- Can be set by clients: no
```
I don't see why we should forbid clients to set this.

majewsky replied:
If this can be set by a client, we have to define how the layout works when fundamental layout properties can change over time. I do not find this a productive use of our time since I don't see a usecase for a client changing this property.

laerling replied:
Accepted.

laerling commented on spec/term/1.0.md:
```diff
@@ -0,0 +1,204 @@
+
+It MAY be computed as follows:
+
+1. For any printable ASCII codepoint `c`, let `N(c)` be the maximum length of a string containing only the codepoint `c`, such that a rendering of this string fits in the width of the terminal viewport without line breaks or overflows.
```
One codepoint? I'm not sure what a codepoint is anymore.

majewsky replied:
Yeah, me neither... :sob:

laerling replied:
No seriously, let's talk about this. I don't see why you calculate a width of a whole (virtual) line by one single codepoint.

majewsky replied:
Will be resolved when this property moves into `mono` since the definition can accordingly be simplified.

laerling commented on spec/mono/1.0.md:
```diff
@@ -0,0 +1,61 @@
+### 1.2. Monospace layouting
+
+We take the term **terminal** to refer to a terminal as defined in [`vt6/term1.0`](https://vt6.io/std/term/1.0/).
+When the value of the `mono.enabled` property is changed, the server SHALL append a paragraph separator (U+2029) to the terminal document, and move the terminal's cursor to point at the end of the terminal document, after that paragraph separator.
```
What's the behavior of form feed compared to paragraph separator? Could using form feed be reasonable as well or even instead?

majewsky replied:
A form feed could trigger pagination on printers. I don't find it helpful to have page breaks between each single shell command and prompt (in the extreme case that the shell does not set `mono.enabled = true`, but every client run by it does).

laerling commented on spec/mono/1.0.md:
```diff
@@ -0,0 +1,61 @@
+
+We take the term **terminal** to refer to a terminal as defined in [`vt6/term1.0`](https://vt6.io/std/term/1.0/).
+When the value of the `mono.enabled` property is changed, the server SHALL append a paragraph separator (U+2029) to the terminal document, and move the terminal's cursor to point at the end of the terminal document, after that paragraph separator.
+After this operation, the terminal's cursor SHALL NOT move past this paragraph separator, even when the specification of another module would allow it.
```
What about text selection by mouse? E. g. when I want to reuse a text snippet of non-mono output again in a new mono-'environment'? Does mouse selection not imply setting the cursor?

majewsky replied:
This specification does not concern a cursor operated by the user. The cursor we're talking about here is denoting the place where additional stdout gets inserted into the terminal document.

laerling commented on spec/mono/1.0.md:
```diff
@@ -0,0 +1,61 @@
+
+*Rationale:* When a client sets `mono.enabled`, prints some text and exits, the shell will then observe the previous value of `mono.enabled` when printing its prompt.
+
+TODO consider upstreaming this scoping logic into core
```
Could it occur that two or more programs inside a single pipe (or just two or more programs that run at the same time on the same frame) compete in a way that e. g. one sets `mono.enabled` to true and expects it to be true until set otherwise, but the other one sets `mono.enabled` to false while the first one still expects it to be true?
This is not only a problem for `mono.enabled`, but is necessary to consider when upstreaming to core.

majewsky replied:
Yes, there is a race condition here, but I don't think it's important in practice. `mono.enabled` defaults to `false`, so most clients that set it are going to set it to true.

laerling replied:
But the race condition could be bad the other way round, as in that a client expects the terminal to be fully unicode-capable while another client already set `mono.enabled` to true a long time ago.

majewsky replied:
After short discussion, the actual problem here is that it should not be possible for a client to set `mono.enabled` *unless* their stdout is connected to the terminal. So this question ties into #19.

majewsky committed:
early draft for vt6/term and vt6/mono

This is based on an earlier draft where monospace layouting was a
capability within vt6/term. I chose to make an entirely new branch
because so much has changed in vt6/core1.0 recently that stuff is way
cleaner this way.

majewsky committed:
adjust for current state of vt6/foundation and vt6/core1.0

majewsky commented:
@laerling I rebased this and adjusted the wording to match #39. There are still plenty of TODOs that I will address at a later time.
